### ISSUE TITLE
[DESIGN] Aria Parquet Optimizations

### ISSUE THREAD

**Issue by yingsu00** [prestodb/presto#16718]
# Background
One year ago FB launched project Aria for ORC readers and greatly improved the Scan performance on FB clusters. In this design document we will introduce a series of similar improvements to the Parquet Reader.

### Moving `TupleDomainFilter` and `FilterFunction` to presto-common
Aria introduced `TupleDomainFilter` interface and `FilterFunction` class and pushed them down to `OrcRecordReader` and `StreamReader` level. Both are in the `com.facebook.presto.orc` package in presto-orc, including most of their implementing classes. On the other hand `TupleDomain`, being an engine internal concept that represents each column to its respective allowable value Domain as a normalized map, is within `com.facebook.presto.common.predicate` package in presto-common. It looks `TupleDomainFilter` and `FilterFunction` are not specific to ORC readers, instead, they are engine concepts that are common to many connectors. Here I propose the following refactoring:

#### TupleDomainFilter interface 
This interface and all its descendants, excluding 
•	`CoercingFilter` in `IntegerNumberUpscaleCoercer`  
•	`CoercingFilter` in `IntegerNumberToVarcharCoercer`

Will be moved 
from `com.facebook.presto.orc` →  `com.facebook.presto.common.predicate`

Specifically, `TrackingTupleDomainFilter` was used in `OrcTester`. The new Parquet reader might need to use it too, so I propose also to move it
from `com.facebook.presto.orc` →  `com.facebook.presto.common.predicate`

#### `FilterFunction` class and related tests 
`com.facebook.presto.orc` →  `com.facebook.presto.common.predicate` 

Its child `TestingFilterFunction` was used in `TestMapFlatSelectiveStreamReader`, but we might need to use it in Parquet reader tests.  Also `TestFilterFunction` doesn't seem to be specific to ORC only. We propose to move all of them from `com.facebook.presto.orc` →  `com.facebook.presto.common.predicate`.

### Consolidate Arrays Utilities
Currently there’re 3 places that defines the commonly used `ensureCapacity()` methods:
•	`com.facebook.presto.orc.array.Arrays`
•	`com.facebook.presto.array.Arrays`
•	`com.facebook.presto.common.block.BlockUtil`

The code was mostly duplicated and scattered in 3 places. The first occurrence was due to a refactoring effort to make presto-orc a standalone library. See (https://github.com/prestodb/presto/issues/14284) and PR https://github.com/prestodb/presto/pull/14418 Drop presto-orc's dependency on presto-array. 

Since these methods would be commonly used by connectors and internal operators, I propose to consolidate the implementation and move them from `com.facebook.presto.array.Arrays` → `com.facebook.presto.common.Arrays`. The other arrays classes like `ShortBigArray`, `AdaptiveLongBigArray` are mostly used by the operators. We can choose not to move them to make presto-orc have minimum dependencies. 

# Design

## Externals
The new Parquet PageSource and readers would be guarded by the following session properties
1.	hive.pushdown_filter_enabled
2.	pushdown_subfields_enabled
3.	hive.parquet_pushdown_filter_enabled

And config properties

1.	hive.pushdown-filter-enabled
2.	experimental.pushdown-subfields-enabled
3.	hive.parquet-pushdown-filter-enabled

Only when both 1, 2 and 3 are true, the new Aria code path would be used.

The first two properties `hive.pushdown_filter_enabled` and `pushdown_subfields_enabled` were introduced in Aria ORC. We will be introducing a new one just for Parquet: `hive.parquet_selective_read_enabled` and `hive.parquet-pushdown-filter-enabled`. This is to provide a switch to disable the new Parquet reader before it’s fully tested for users who read both ORC and Parquet files. The Parquet specific property will be removed after the new reader proved to be stable and performant.

## Parquet rowgroup skipping
Parquet file has file level metadata and rowgroup level metadata and both are stored in the file footer:
 
![image](https://user-images.githubusercontent.com/33299678/132636564-133c56f8-5205-4afb-9ad1-0dce565638c2.png)

These map to the following classes and fields:

- ParquetMetadata
  - FileMetaData
  - BlockMetaData (one for each rowgroup)
    - ColumnChunkMetaData (one for each column each rowgroup)
      - valueCount
      - totalSize
      - totalUncompressedSize
      - Statistics
        - Min 
        - max 
        - numNulls 

We will create a new `ParquetSelectivePageSourceFactory` and `ParquetSelectivePageSource` for the Aria Parquet optimizations. 

Based on the statistics in `ColumnChunkMetaData`, the new `ParquetSelectivePageSourceFactory` would eliminate unwanted rowgroups while constructing the new `ParquetSelectivePageSource`. Only the rowgroups passing the domain predicate would be collected and stored in the page source for later reading

## Filter Pushdown
The filters pushed down to the Page source are two types:
•	TupleDomain filter
•	Remaining filter

The TupleDomain filter is a kind of filter that only involves one columns and can be expressed as value ranges. The remaining filters would be converted to filter functions. These filters are read from the Hive metastore and passed to the Parquet page source. The process of pushing down the filters were already done in Aria ORC project. We just need to add the new `ParquetSelectivePageSourceFactory` to the `selectivePageSourceFactoryBinder` in `HiveMetadata` and enable the tests for pushdown_filter_enabled when the Hive format is Parquet.

#### Overview

We will introduce a new `ParquetSelectiveReader` which manages a set of new selective column readers, one for each column. The `ParquetSelectiveReader` maintains a `positions` array. Initially the array contains all positions in the batch. While the column readers are reading the data, the `TupleDomainFilter`s on this column will be evaluated during the reading and decoding process. The remaining positions will be passed to the next column reader. If rows are filtered out by the previous column readers, these rows would be skipped and not materialized for the current column. 

If a column has both `TupleDomainFilter`s and `FilterFunction`s, the `FilterFunction`s would be evaluated after the `TupleDomainFilter`s. The `FilterFunction` is applied on a `Page`. To avoid data copying before the finalization phase, we create a Page by wrapping data values in the corresponding column readers using BlockLease. 

When all filters are evaluated, the data at the final qualifying row numbers are extracted and compacted to form the output pages and blocks.

#### Null handling

The Parquet file format uses definition level to identify nulls. https://blog.twitter.com/engineering/en_us/a/2013/dremel-made-simple-with-parquet is a good read to understand the concept. Like other integer types, it’s encoded using RLE, PACKED encodings. The existing `DefinitionLevelDecoder` will be used to decode the definition level values. For each position, if its definition level value is less than the max definition level, the position is null.

Note that we should avoid reading nulls unless the num_nulls in `ColumnChunkMetaData` statistics is greater than 0. This was not implemented in existing Parquet readers.

#### Skip Mode

We introduce two modes in the column readers. The first mode is “skip mode”. In this mode, we first seek to the current position, then decode and read the position and go to the next position, until all positions are read. The seek will try not do decode the data unless it’s absolutely required. The whole page would be skipped if there is no rows in this page to be read. 

When there’re nulls, we would also do skipping on the definition levels in a similar way of reading values. 

This mode works well for very highly selective filters.

#### Batch Read Mode

While the skip mode works for very sparse positions, sometimes it’s more efficient to read all the values in batch. For example, when there are no nulls, or filter rate is low, this mode works faster than the skip mode. The idea is similar to the current Parquet batch readers, but we will evaluate the filters before the Page and Blocks are materialized. By evaluating the filters in a tight loop, it will also be more efficient than evaluating it value by value. If there are nulls, we need one more step before applying the filters: we need to unpack the values according to the nulls. 

In the final materialization step when we’re creating the `Block`s, the values would be packed according to the final positions. When there is no filter functions , we can choose to do the packing step at the very end of reading all columns, not right after reading each column. This is to avoid the extra memory copy by just packing the data once. 

### Array, Map and Row

The subfield pruning planner change was already implemented in Aria ORC. We will just additionally check if `parquet_selective_read_enabled` is enabled. The fields in a complex type column will be expressed as the existing `Subfield` class in presto-common.

Reading a complex type follows a similar process as the primitive types. The leaf level fields would be extracted and used to create the column reader. We will read the repetition levels for each value to identify which nested level it is in, and translate this information into offsets. 

There will also be a “skip mode” and a “batch mode”. In both modes we need to read an array of positions from the top level column. We populate the positions for the nested level based on the offsets we created before. Then based on selectivity and null information, we pick one mode. 

## Caching

This includes metadata caching and data caching. The meta data layout is different from ORC so the code needs to be modified. 
To use Alluxio data caching, we will replace the current Hadoop file system to `AlluxioFileSystem`. The `ParquetDataSource` will be using the `AlluxioHdfsInputStream`.

We will work on caching after the previous sections are implemented.

# Implementation Principles

We want to achieve very good performance and will follow several principles

## Use tight loops
We need to optimize performance critical loops, so that the compiler can auto-vectorize. Avoid branching within loops when possible. Sometimes we need to separate the loops to optimize certain scenarios, e.g. the no null case.

## Avoid using dynamic containers
The dynamic arrays like `IntArrays` usually have worse performance than primitive type arrays like `int[]`. They can be used on non performance critical structures and the number of values is not pre-known. In our case we know the number of values to read and and pre-allocate and grow the arrays. 

## Avoid memory allocations and copies
Memory allocation is very important for high performance computing. Excessive allocations can also trigger reliability issues. We will try to reuse all the arrays and large data structures whenever possible.

Memory copy should also be avoided as much as possible. For example, when there is no filter functions, we can do compaction only once for all columns so that we won’t copy the data twice.

# Further Optimizations

## Compression Skipping or Indexing
Most compression algorithms has the concept of “segments” or “members”. Usually the before and after data sizes are kept in the segment header. For example a gzip file consists of a series of "members" (compressed data sets).  The format of each member is specified in the following section.  The members simply appear one after another in the file, with no additional information before, between, or after them.
https://datatracker.ietf.org/doc/html/rfc1952#page-5 

Theoretically we should be able to skip the segments if there are no positions to read in them. This works on fix-width types like INT, BIGINT etc. However we will need to extend the existing Java API to make it support skipping.

## Pushdown coercer application to `ColumnReader`

The existing coercers uses `BlockBuilder`s and is not very efficient. Plus, it requires two memory copies: the first one creates the `Block`, the second one applies the coercer. We can push down the coercing into the `Page`/`Block` materialization step, which only needs one memory copy, and avoid the expensive building process.

## Filter Evaluation on Encoded Data
Currently in all Presto code path, we needed to decode the data before applying the filters. However in many cases we can apply the filter on the RLE or PACKED data directly. This avoids poplulating the data to the values array. More importantly, we can apply the filter on multiple values at once. For example, if values are 4-bits packed, we can evaluate filter on 16 values using one instruction. Only when the result is not non-match we need to extract the position and value. If we can use SIMD intrinsics, it will improve the performance a lot.

## ColumnIndex Support

In latest version of Parquet https://github.com/apache/parquet-format there is a new form of meta data called “ColumnIndex”: 

We add two new per-column structures to the row group metadata:
•	ColumnIndex: this allows navigation to the pages of a column based on column values and is used to locate data pages that contain matching values for a scan predicate
•	OffsetIndex: this allows navigation by row index and is used to retrieve values for rows identified as matches via the ColumnIndex. Once rows of a column are skipped, the corresponding rows in the other columns have to be skipped. Hence the OffsetIndexes for each column in a RowGroup are stored together.
![image](https://user-images.githubusercontent.com/33299678/132636632-e47e78d7-2bca-472f-813c-1a159e39ba03.png)

 
The new index structures are stored separately from RowGroup, near the footer.

Before this is added, we weren’t able to skip the pages unless we read the page header. With this ColumnIndex, we can read the metadata in one IO and pre-eliminate the positions to read.

The current reader and writer both need to support this. If ColumnIndex is unavailable it would fall back to the process we described earlier.

## Bloom Filter Support

The latest version of Parquet format also supports Bloom filters: https://github.com/apache/parquet-format/blob/3fb10e00c2204bf1c6cc91e094c59e84cefcee33/BloomFilter.md 
![image](https://user-images.githubusercontent.com/33299678/132636649-692b4b23-7729-40c4-84a0-ce5f37b3d445.png)

 
The writer can choose to create the Bloom filter index if 1) number of distinct values is not big 2) history shows this column may have heavy use of equality filters in the workload.

## Integration with secondary indexes
In the future Presto may support native secondary indexes like B+ tree index or inverted index to support faster lookups. The access method would need to decide if reading from the index would help the reader to achieve higher performance. This is similar to filter reordering: we will compare the cost and selectivity of retrieving the destination records with directly reading from the parquet files with filters. If it helps reduce the majority of rows, we can set the start positions array to be the ones gotten from the secondary index. The rest process doesn’t change.

**Comment by yingsu00** [comment]
cc @zhenxiao @mbasmanova @apc999 @ClarenceThreepwood @simmend @biswapesh

**Comment by yingsu00** [comment]
this is not done yet. 
WIP PR https://github.com/prestodb/presto/pull/17085

**Comment by shangxinli** [comment]
ColumnIndex for reader is in this [PR](https://github.com/prestodb/presto/pull/16963)
